File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_101_200/Sol101.java
package com.goinhn.thorwcode.leetcode.question._101_200;

/**
 * 101.对称二叉树
 *
 * @author goinhn
 * @date 2021/11/20
 */
public class Sol101 {

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    /**
     * 广度优先搜索，转换为两个相同的树问题
     * <p>
     * time:O(min(m,n))
     * space:O(min(m,n))
     */
    private static class Solution1 {
        public boolean isSymmetric(TreeNode root) {
            if (root == null) {
                return true;
            }
            return isSame(root.left, root.right);
        }

        private boolean isSame(TreeNode leftTree, TreeNode rightTree) {
            if (leftTree == null && rightTree == null) {
                return true;
            } else if (leftTree == null || rightTree == null) {
                return false;
            } else if (leftTree.val != rightTree.val) {
                return false;
            } else {
                return isSame(leftTree.left, rightTree.right) && isSame(leftTree.right, rightTree.left);
            }
        }
    }
}

File: leetcode_go/question/_101_200/_101/sol_101.go
// Package _101 对称二叉树
// @author goinhn
// @data 2022/8/20
package _101

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 深度优先搜索
// time:O(n)
// space:O(n)
func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return dfs(root.Left, root.Right)
}

func dfs(leftNode *TreeNode, rightNode *TreeNode) bool {
	if leftNode == nil && rightNode == nil {
		return true
	}

	if leftNode == nil || rightNode == nil {
		return false
	}

	if leftNode.Val != rightNode.Val {
		return false
	} else {
		// 左右两侧相互比较
		return dfs(leftNode.Left, rightNode.Right) && dfs(leftNode.Right, rightNode.Left)
	}
}
